### PR TITLE
LITE-21943 - Add support for python’s list of dict filtering.

### PR DESCRIPTION
**Pull request by ghost**
Add support for python’s list of dict filtering.

### CODE REVIEW

**Review by marcserrat — APPROVED** [review]
thanks for also making it compatible with legacy pythons and align with django-rql!

File: py_rql/exceptions.py [hunk 1]
```diff
@@ -29,3 +29,11 @@ class RQLFilterLookupError(RQLFilterError):
 class RQLFilterValueError(RQLFilterError):
     """ Value error is raised when provided values can't be converted to DB field types. """
     MESSAGE = 'RQL Value error.'
+
+
+class RQLFilterNotFoundError(RQLFilterError):
```
Comment (maxipavlovic) [inline py_rql/exceptions.py]: This error is doubtful for me as the original idea was to simply skip non described filters (this is how Django-RQL works, for example). I would suggest to keep following "open" approach and, thus, would suggest to remove this error type.

File: py_rql/filter.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+#
```
Comment (maxipavlovic) [inline py_rql/filter.py]: I don't like shadowing of builtin names, so I would propose to rename file to `filter_cls.py` by analogy with Django-RQL.

File: py_rql/filter.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    FILTERS = []
+
+    def __init__(self):
+        self._cache = LFUCache(maxsize=1000)
+        self._filters = self.init_filters()
+
+    def init_filters(self):
+        filters = {}
+        for filter_decl in self.FILTERS:
```
Comment (maxipavlovic) [inline py_rql/filter.py]: Please, insert blank lines between logically separated pieces of codes everywhere. Currently, it's extremely hard for me to read. Just for example:
```python
            filter_name = filter_decl['filter']
            e = f"'{filter_name}' is a reserved filter name."
            assert filter_name not in RESERVED_FILTER_NAMES, e

            filter_type = filter_decl.get('type', FilterTypes.STRING)
            default_lookups = self.default_filter_lookups_for_type(filter_type)
            lookups = filter_decl.get('lookups', default_lookups)
            e = f"Invalid lookups for filter '{filter_name}' of type '{filter_type}': '{lookups}'"
            assert set(lookups).issubset(default_lookups), e

            cast_func = filter_decl.get(
                'cast_func',
                get_default_cast_func_for_type(filter_type),
            )
            e = f"Invalid cast function for filter '{filter_name}'."
            assert callable(cast_func), e

            filters[filter_name] = {
                'type': filter_type,
                'lookups': lookups,
                'cast_func': cast_func,
            }
....
```

File: py_rql/parser.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 #
 
 from cachetools import LFUCache
-
```
Comment (maxipavlovic) [inline py_rql/parser.py:6]: ?)

File: py_rql/filter.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+from py_rql.cast import get_default_cast_func_for_type
+from py_rql.constants import FilterLookups, FilterTypes, RESERVED_FILTER_NAMES
+from py_rql.exceptions import (
+    RQLFilterLookupError,
+    RQLFilterNotFoundError,
+)
+from py_rql.transformer import RQLToFunctionTransformer
+
+
+class Filter:
```
Comment (maxipavlovic) [inline py_rql/filter.py]: I don't like to discuss naming, but here I think that it's important. I had the same issue with Django-RQL.
I would suggest `FilterClass`. Please, ping me in Zoom for me to provide the reasoning.

File: README.md [hunk 6]
```diff
@@ -1,27 +1,55 @@
-Python RQL
```
Comment (maxipavlovic) [inline README.md:1]: Few notes on README:
1. I don't like dependencies described in it as we have poetry and we need to have 2 places to keep in sync :)
2. Some rendering error with ##
3. Simplest example of usage of filter class is totally needed in README
4. Link on Downloads shields is wrong and should be: (https://pypi.org/project/lib-rql/)

image

**Review by maxipavlovic — CHANGES_REQUESTED** [review]
Incredible job @ffaraoneim !

File: README.md (hunk 6 above)
Comment (marcserrat) [reply README.md:1]: i suggest that documentation we go over readthedocs as is being prepared.......way better in my pov.

**Review by marcserrat — APPROVED** [review]

File: examples/books.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+from pprint import pprint
+
+from py_rql.constants import FilterTypes
+from py_rql.filter_cls import Filter
+
+
+DATA = json.load(open('./books.json'))['books']
+
+
+class BookFilter(Filter):
```
Comment (maxipavlovic) [inline examples/books.py]: It's not Filter now :)

**Review by maxipavlovic — CHANGES_REQUESTED** [review]

File: examples/books.py (hunk 7 above)
Comment (ghost) [reply examples/books.py]: Thanks!

**Review by maxipavlovic — APPROVED** [review]

**Pull request merged**